File: creational/factory-method/document_factory.py
from abc import ABC, abstractmethod

# Product interface
class Document(ABC):
    @abstractmethod
    def create(self):
        pass

# Concrete Products
class PDFDocument(Document):
    def create(self):
        return "PDF document created"

class WordDocument(Document):
    def create(self):
        return "Word document created"

class TextDocument(Document):
    def create(self):
        return "Text document created"

# Creator (Factory) class
class DocumentCreator:
    @staticmethod
    def create_document(doc_type):
        """Factory method to create documents based on type."""
        if doc_type == "pdf":
            return PDFDocument()
        elif doc_type == "word":
            return WordDocument()
        elif doc_type == "text":
            return TextDocument()
        else:
            raise ValueError(f"Document type '{doc_type}' not supported")


if __name__ == "__main__":
    # Create different document types using the factory
    for doc_type in ["pdf", "word", "text"]:
        document = DocumentCreator.create_document(doc_type)
        print(document.create())
    
    # Try with an invalid type (will raise an exception)
    try:
        document = DocumentCreator.create_document("excel")
        print(document.create())
    except ValueError as e:
        print(f"Error: {e}")

File: creational/factory-method/DocumentFactory.java
// Product interface
interface Document {
    String create();
}

// Concrete Products
class PDFDocument implements Document {
    @Override
    public String create() {
        return "PDF document created";
    }
}

class WordDocument implements Document {
    @Override
    public String create() {
        return "Word document created";
    }
}

class TextDocument implements Document {
    @Override
    public String create() {
        return "Text document created";
    }
}

// Creator (Factory) class
class DocumentCreator {
    public static Document createDocument(String docType) {
        // Factory method to create documents based on type
        switch (docType.toLowerCase()) {
            case "pdf":
                return new PDFDocument();
            case "word":
                return new WordDocument();
            case "text":
                return new TextDocument();
            default:
                throw new IllegalArgumentException("Document type '" + docType + "' not supported");
        }
    }
}

public class DocumentFactory {
    public static void main(String[] args) {
        // Create different document types using the factory
        String[] docTypes = {"pdf", "word", "text"};
        
        for (String type : docTypes) {
            Document document = DocumentCreator.createDocument(type);
            System.out.println(document.create());
        }
        
        // Try with an invalid type (will throw an exception)
        try {
            Document document = DocumentCreator.createDocument("excel");
            System.out.println(document.create());
        } catch (IllegalArgumentException e) {
            System.out.println("Error: " + e.getMessage());
        }
    }
}